source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,2515 @@
+  response). If false, the test runner MUST assert that the error does not
+  originate from the client.
+
+  Client errors include, but are not limited to: parameter validation errors
+  before a command is sent to the server; network errors.
+
+- ``errorContains``: Optional string. A substring of the expected error message
+  (e.g. "errmsg" field in a server error document). The test runner MUST assert
+  that the error message contains this string using a case-insensitive match.
+
+  See `bulkWrite`_ for special considerations for BulkWriteExceptions.
+
+- ``errorCode``: Optional integer. The expected "code" field in the
+  server-generated error response. The test runner MUST assert that the error
+  includes a server-generated response whose "code" field equals this value.
+  In the interest of readability, YAML files SHOULD use a comment to note the
+  corresponding code name (e.g. ``errorCode: 26 # NamespaceNotFound``).
+
+  Server error codes are defined in
+  `error_codes.yml <https://github.com/mongodb/mongo/blob/master/src/mongo/base/error_codes.yml>`__.
+
+  Test files SHOULD NOT assert error codes for client errors, as specifications
+  do not define standardized codes for client errors.
+
+- ``errorCodeName``: Optional string. The expected "codeName" field in the
+  server-generated error response. The test runner MUST assert that the error
+  includes a server-generated response whose "codeName" field equals this value
+  using a case-insensitive comparison.
+
+  See `bulkWrite`_ for special considerations for BulkWriteExceptions.
+
+  Server error codes are defined in
+  `error_codes.yml <https://github.com/mongodb/mongo/blob/master/src/mongo/base/error_codes.yml>`__.
+
+  Test files SHOULD NOT assert error codes for client errors, as specifications
+  do not define standardized codes for client errors.
+
+- ``errorLabelsContain``: Optional array of strings. A list of error label
+  strings that the error is expected to have. The test runner MUST assert that
+  the error contains all of the specified labels (e.g. using the
+  ``hasErrorLabel`` method).
+
+- ``errorLabelsOmit``: Optional array of strings. A list of error label strings
+  that the error is expected not to have. The test runner MUST assert that
+  the error does not contain any of the specified labels (e.g. using the
+  ``hasErrorLabel`` method).
+
+.. _expectedError_expectedResult:
+
+- ``expectedResult``: Optional mixed type. This field follows the same rules as
+  `operation.expectedResult <operation_expectedResult_>`_ and is only used in
+  cases where the error includes a result (e.g. `bulkWrite`_). If specified, the
+  test runner MUST assert that the error includes a result and that it matches
+  this value.
+
+
+expectedEvent
+~~~~~~~~~~~~~
+
+An event (e.g. APM), which is expected to be observed while executing the test's
+operations.
+
+This document MUST contain **exactly one** top-level key that identifies the
```

Comment:
In a couple places in the spec it talks about arrays of objects, that contain
one top-level property / key. 

I suggest that this key be a property of the object, and not to have this
behavior.

My stance on this comes from the nested nature of objects and arrays.

Both options, 1) single keyed objects 2) objects where "key" is a property, seem
to be viable solutions. 

I see a key as a limiting factor where a object can only have one. In these
cases because it's an array, there can be multiple, so it's not relevant to have
it as a key.

My proposal:

```js
{
  observeEvents: [
    {
      name: 'commandStartedEvent',
      command: '42',
      databaseName: 'forty-two'
    }
  ]
}
```

Current design:

```js
{
  observeEvents: [
    {
      commandStartedEvent: {
        command: '42',
        databaseName: 'forty-two'
      }
    }
  ]
}
```

This I've already been a proponent of changing above in entities, but I would
say this would be my suggestion for any of these cases.

I simply think it lowers the complexity and makes the conditional nature of the
runner easier to use. The condition will need to be matched regardless,
however when my suggestion we have less spec validation for having to have
single keyed objects.

It also allows better typing to have a single "nugget" / object, that does not
depend on it's parent key as reference.

Reply:
Will leave this open pending outcome of the similar proposal for elements in `createEntities`. Provided we can make the JSON schema validation work easily, I'm open to revising this.

Reply:
Not sure if this is true for other languages, but at least for Rust, the single-key representation allows for easier translations between JSON and our data types. For example, we have a `CommandStartedEvent` struct, so when parsing the JSON files, Serde (the Rust library we use to translate between JSON/BSON and Rust structs/enums) would recognize that the single key corresponds to the struct name and properly serialize it as a `CommandStartedEvent` -- eliminating this key would require manually parsing out the name key instead. For this reason, I find the single-key representation to be a little more intuitive (although it's possible that it would go in the other direction for languages that parse differently). It also better corresponds with the [API for command monitoring](https://github.com/mongodb/specifications/blob/master/source/command-monitoring/command-monitoring.rst#api) (which doesn't embed the type of the event).

Reply:
I concur that the single-key wrapping makes parsing in PHP easier as well. When we receive the parsed JSON as a generic object, we can inspect that field and then delegate to a particular class to construct itself from the value (i.e. object for the entity or event). I'd argue this also makes the test files more readable, as you can scan a column and quickly see a sequence of types apart from the rest of the object body.

Reply:
Now that I've implemented single-key wrapping It's not as painful as I once thought it would be, and @jmikola's solution for validating using `{ "minProperties": 1, "maxProperties": 1 }` sounds great. 

@isabelatkinson  I also agree it makes typings / accessibility easier in some regards.